<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>InteRD</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">InteRD</h1>



<div id="installation" class="section level2">
<h2>Installation</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">&quot;devtools&quot;</span>)) {</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;chencxxy28/InteRD&quot;</span>)</span></code></pre></div>
<p>In this tutorial, we use SCDC (<a href="https://academic.oup.com/bib/article/22/1/416/5699815">Dong et
al.Â (2021)</a>) for illustration to conduct reference-based
deconvolution.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#load</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(InteRD)</span></code></pre></div>
</div>
<div id="implementation-case-study-1" class="section level2">
<h2>Implementation (Case study 1)</h2>
<p>First, read in scRNA-seq data from our website</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>readRDSFromWeb <span class="ot">&lt;-</span> <span class="cf">function</span>(ref) {</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">readRDS</span>(<span class="fu">gzcon</span>(<span class="fu">url</span>(ref)))</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>seger <span class="ot">&lt;-</span> <span class="fu">readRDSFromWeb</span>(<span class="st">&quot;https://github.com/chencxxy28/Data/raw/main/data_InteRD/segerstolpe.rds&quot;</span>)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>baron <span class="ot">&lt;-</span> <span class="fu">readRDSFromWeb</span>(<span class="st">&quot;https://github.com/chencxxy28/Data/raw/main/data_InteRD/baron.rds&quot;</span>)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>xin<span class="ot">&lt;-</span><span class="fu">readRDSFromWeb</span>(<span class="st">&quot;https://github.com/chencxxy28/Data/raw/main/data_InteRD/Xin_nonD.rds&quot;</span>)</span></code></pre></div>
<p>Second, use function <code>generateBulk()</code> to create the
<em>pseudo bulk</em> samples, where <code>ct.varname</code> specifies
the name of cell type clustering result variable. <code>sample</code>
specifies the name of subjects information variable. <code>ct.sub</code>
specifies the names of cell types used to construct <em>pseudo bulk</em>
samples. Here we provide an example where the <em>pseudo bulk</em>
samples are generated from <a href="https://www.sciencedirect.com/science/article/pii/S1550413116304363?via%3Dihub">Segerstolpe
et al.Â (2016)</a>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234567</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>pseudo.seger<span class="ot">&lt;-</span><span class="fu">generateBulk</span>(seger[[<span class="st">&quot;sc.eset.qc&quot;</span>]], <span class="at">ct.varname =</span> <span class="st">&quot;cluster&quot;</span>, <span class="at">sample =</span> <span class="st">&quot;sample&quot;</span>, <span class="at">ct.sub =</span> <span class="fu">c</span>(<span class="st">&quot;alpha&quot;</span>,<span class="st">&quot;beta&quot;</span>,<span class="st">&quot;delta&quot;</span>,<span class="st">&quot;gamma&quot;</span>), <span class="at">nbulk =</span> <span class="dv">40</span>, <span class="at">low_s =</span> <span class="fl">0.3</span>, <span class="at">upp_s =</span> <span class="fl">0.7</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>truep<span class="ot">&lt;-</span>pseudo.seger<span class="sc">$</span>true_p[<span class="fu">complete.cases</span>(pseudo.seger<span class="sc">$</span>true_p),]</span></code></pre></div>
<p>The generated <em>pseudo bulk</em> object contains a matrix of true
cell type proportions (<code>pseudo.seger$truep</code>) and the
<code>ExpressionSet</code> object
(<code>pseudo.seger$pseudo_eset</code>).</p>
<p>InteRD1 algorithm is the first step of InteRD to do multiple
reference ensemble, which takes as input the target bulk data, a list of
marker genes (pre-selected by <code>Seurat</code> based on a single cell
data from <a href="https://www.sciencedirect.com/science/article/pii/S155041311630434X?via%3Dihub">Xin
et al.Â (2016)</a>), a list of subject-level cell type proportions based
on each reference set. For illustration, we use the function
<code>SCDC_ENSEMBLE()</code> from SCDC package (<a href="https://academic.oup.com/bib/article/22/1/416/5699815">Dong et
al.Â (2021)</a>) to obtain single-reference-based estimates and ensemble
estimates. For reference panels, we consider two single cell data, one
from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5228327/">Baron et
al.Â (2016)</a>, and the other from <a href="https://www.sciencedirect.com/science/article/pii/S155041311630434X?via%3Dihub">Xin
et al.Â (2016)</a>. The estimates can be obtained by running the
following code</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234567</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SCDC)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="do">##ensemble of multiple reference sets </span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co">#resuts based on SCDC</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  pancreas.sc <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">baron =</span> baron<span class="sc">$</span>sc.eset.qc,</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>                      <span class="at">xin   =</span> xin</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>  SCDC_results<span class="ot">&lt;-</span><span class="fu">SCDC_ENSEMBLE</span>(<span class="at">bulk.eset =</span> pseudo.seger<span class="sc">$</span>pseudo_eset, <span class="at">sc.eset.list =</span> pancreas.sc, <span class="at">ct.varname =</span> <span class="st">&quot;cluster&quot;</span>,</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>                                <span class="at">sample =</span> <span class="st">&quot;sample&quot;</span>, <span class="at">weight.basis =</span><span class="cn">TRUE</span>,<span class="at">truep =</span> truep, <span class="at">ct.sub =</span>  <span class="fu">c</span>(<span class="st">&quot;alpha&quot;</span>,<span class="st">&quot;beta&quot;</span>,<span class="st">&quot;delta&quot;</span>,<span class="st">&quot;gamma&quot;</span>), <span class="at">search.length =</span> <span class="fl">0.02</span>,<span class="at">grid.search=</span><span class="cn">TRUE</span>)</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>  comb<span class="ot">&lt;-</span>SCDC_results<span class="sc">$</span>prop.only</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>  weight_matrix<span class="ot">&lt;-</span>SCDC_results<span class="sc">$</span>w_table[<span class="st">&quot;mAD_Y&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>  SCDC_ENSEMBLE_MAD<span class="ot">&lt;-</span>SCDC<span class="sc">:::</span><span class="fu">wt_prop</span>(weight_matrix,comb)</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># saveRDS(SCDC_ENSEMBLE_MAD,&quot;/Users/chixiang.chen/Library/CloudStorage/OneDrive-UniversityofMarylandSchoolofMedicine/postdoc/postdoc/deconvolution/ref_based_rd/data_InteRD/SCDC_ENSEMBLE_MAD_seger.rds&quot;)</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># saveRDS(comb,&quot;/Users/chixiang.chen/Library/CloudStorage/OneDrive-UniversityofMarylandSchoolofMedicine/postdoc/postdoc/deconvolution/ref_based_rd/data_InteRD/comb_seger.rds&quot;)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">#results based on InteRD1</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>SCDC_ENSEMBLE_MAD<span class="ot">&lt;-</span><span class="fu">readRDSFromWeb</span>(<span class="st">&quot;https://github.com/chencxxy28/Data/raw/main/data_InteRD/SCDC_ENSEMBLE_MAD_seger.rds&quot;</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>comb<span class="ot">&lt;-</span><span class="fu">readRDSFromWeb</span>(<span class="st">&quot;https://github.com/chencxxy28/Data/raw/main/data_InteRD/comb_seger.rds&quot;</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  list_marker<span class="ot">&lt;-</span><span class="fu">readRDSFromWeb</span>(<span class="st">&quot;https://github.com/chencxxy28/Data/raw/main/data_InteRD/list_markerbaron20.rds&quot;</span>) <span class="co">#get markers selected from xin et al (2016)</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  lambda_option<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">0.01</span>,<span class="fl">0.05</span>,<span class="fl">0.1</span>,<span class="dv">1</span>,<span class="dv">5</span>,<span class="dv">100</span>)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>  cell_type_unique<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">&quot;alpha&quot;</span>,<span class="st">&quot;beta&quot;</span>,<span class="st">&quot;delta&quot;</span>,<span class="st">&quot;gamma&quot;</span>)</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  InteRD1.output<span class="ot">&lt;-</span><span class="fu">InteRD1</span>(<span class="at">bulk.data =</span>pseudo.seger<span class="sc">$</span>pseudo_eset,list_marker,cell_type_unique,<span class="at">comb_used=</span>comb,lambda_option)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>  InteRD1<span class="ot">&lt;-</span><span class="fu">InteRD.predict.prop</span>(<span class="at">InteRD.output=</span>InteRD1.output)</span></code></pre></div>
<p>We provide the function <code>evaluate</code> to assess the
performance of estimated proportions versus the true proportions based
on mean absolute deviance (MAD, the smaller the better), Kendall
correlation coefficient (Ken, the larger the better), and Pearson
correlation coefficient (Cor, the larger the better).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">evaluate</span>(SCDC_ENSEMBLE_MAD,pseudo.seger<span class="sc">$</span>true_p)<span class="sc">$</span>all.eva</span></code></pre></div>
<pre><code>##     all.mad   all.ken   all.cor
## 1 0.1069572 0.7358974 0.8939433</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">evaluate</span>(InteRD1,pseudo.seger<span class="sc">$</span>true_p)<span class="sc">$</span>all.eva</span></code></pre></div>
<pre><code>##      all.mad     all.ken    all.cor
## 1 0.04851359 0.004487179 0.02931761</code></pre>
<p>InteRD2 algorithm is the second step of InteRD to further integrate
prior biological information into the deconvolution in a robust manner.
The prior information is the population-level mean cell-type proportions
and their corresponding standard deviations across samples. In this
tutorial, we obtain this information from scRNA-seq data from <a href="https://www.sciencedirect.com/science/article/pii/S1550413116304363?via%3Dihub">Segerstolpe
et al.Â (2016)</a>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>ave_est <span class="ot">=</span> <span class="fu">pop.ct.prop.scRNA</span>(<span class="at">scRNA=</span>seger[[<span class="st">&quot;sc.eset.qc&quot;</span>]],<span class="at">cell_type_unique=</span>cell_type_unique)<span class="sc">$</span>pop.ct.prop</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>ave_sd <span class="ot">=</span> <span class="fu">pop.ct.prop.scRNA</span>(<span class="at">scRNA=</span>seger[[<span class="st">&quot;sc.eset.qc&quot;</span>]],<span class="at">cell_type_unique=</span>cell_type_unique)<span class="sc">$</span>pop.ct.sd</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>lambda_option<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="fu">seq</span>(<span class="at">from=</span><span class="dv">1</span>,<span class="at">to=</span><span class="dv">20</span>,<span class="at">length=</span><span class="dv">4</span>),<span class="fu">seq</span>(<span class="at">from=</span><span class="dv">30</span>,<span class="at">to=</span><span class="dv">100</span>,<span class="at">length=</span><span class="dv">4</span>),<span class="dv">200</span>,<span class="dv">500</span>,<span class="dv">1000000</span><span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>InteRD2.output<span class="ot">&lt;-</span><span class="fu">InteRD2</span>(<span class="at">bulk.data=</span>pseudo.seger<span class="sc">$</span>pseudo_eset,list_marker,cell_type_unique,<span class="at">comb_sampled=</span>InteRD1,ave_est,ave_sd,<span class="at">lambda_option=</span>lambda_option)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>InteRD2<span class="ot">&lt;-</span><span class="fu">InteRD.predict.prop</span>(<span class="at">InteRD.output=</span>InteRD2.output)</span></code></pre></div>
<p>Note that Reference-free approach and TOAST are special cases of
InteRD2. The TOAST can be run based on the function <code>InteRD2</code>
by specifying <code>lambda_option=0</code>, and Reference-free approach
can be run by the function <code>Ref_free</code>. The following is the
demo for reference-free approach.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>ref_free.output<span class="ot">&lt;-</span><span class="fu">Ref_free</span>(<span class="at">bulk.data=</span>pseudo.seger<span class="sc">$</span>pseudo_eset,<span class="at">list_marker=</span>list_marker,<span class="at">cell_type_unique=</span>cell_type_unique)</span></code></pre></div>
<pre><code>## [1] 1</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>reffree<span class="ot">&lt;-</span><span class="fu">InteRD.predict.prop</span>(<span class="at">InteRD.output=</span>ref_free.output)</span></code></pre></div>
<p>Based on the evaluations, we can tell that InteRD2 is better than
InteRD1, the InteRD estimates are all better than the existing
method.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">evaluate</span>(SCDC_ENSEMBLE_MAD,pseudo.seger<span class="sc">$</span>true_p)<span class="sc">$</span>all.eva</span></code></pre></div>
<pre><code>##     all.mad   all.ken   all.cor
## 1 0.1069572 0.7358974 0.8939433</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">evaluate</span>(reffree,pseudo.seger<span class="sc">$</span>true_p)<span class="sc">$</span>all.eva</span></code></pre></div>
<pre><code>##      all.mad   all.ken   all.cor
## 1 0.03773585 0.7487179 0.9264559</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">evaluate</span>(InteRD1,pseudo.seger<span class="sc">$</span>true_p)<span class="sc">$</span>all.eva</span></code></pre></div>
<pre><code>##      all.mad     all.ken    all.cor
## 1 0.04851359 0.004487179 0.02931761</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">evaluate</span>(InteRD2,pseudo.seger<span class="sc">$</span>true_p)<span class="sc">$</span>all.eva</span></code></pre></div>
<pre><code>##      all.mad   all.ken   all.cor
## 1 0.01548666 0.7583333 0.9313362</code></pre>
</div>
<div id="implementation-case-study-2" class="section level2">
<h2>Implementation (Case study 2)</h2>
<p>Now let us consider another pseudo data generated from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5228327/">Baron et
al.Â (2016)</a>.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234567</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>pseudo.baron<span class="ot">&lt;-</span><span class="fu">generateBulk</span>(baron[[<span class="st">&quot;sc.eset.qc&quot;</span>]], <span class="at">ct.varname =</span> <span class="st">&quot;cluster&quot;</span>, <span class="at">sample =</span> <span class="st">&quot;sample&quot;</span>, <span class="at">ct.sub =</span> <span class="fu">c</span>(<span class="st">&quot;alpha&quot;</span>,<span class="st">&quot;beta&quot;</span>,<span class="st">&quot;delta&quot;</span>,<span class="st">&quot;gamma&quot;</span>), <span class="at">nbulk =</span> <span class="dv">40</span>, <span class="at">low_s =</span> <span class="fl">0.3</span>, <span class="at">upp_s =</span> <span class="fl">0.7</span>)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>truep<span class="ot">&lt;-</span>pseudo.baron<span class="sc">$</span>true_p[<span class="fu">complete.cases</span>(pseudo.baron<span class="sc">$</span>true_p),]</span></code></pre></div>
<p>Then we apply SCDC to obtain the estimates based on InteRD and
SCDC-ENSEMBLE, where we consider two single cell data, one from <a href="https://www.sciencedirect.com/science/article/pii/S1550413116304363?via%3Dihub">Segerstolpe
et al.Â (2016)</a>, and the other from <a href="https://www.sciencedirect.com/science/article/pii/S155041311630434X?via%3Dihub">Xin
et al.Â (2016)</a>. The estimates can be obtained by running the
following code</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234567</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="do">##ensemble of multiple reference sets </span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co">#resuts based on SCDC</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  pancreas.sc <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">seger =</span> seger<span class="sc">$</span>sc.eset.qc,</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">xin   =</span> xin</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>  SCDC_results<span class="ot">&lt;-</span><span class="fu">SCDC_ENSEMBLE</span>(<span class="at">bulk.eset =</span> pseudo.baron<span class="sc">$</span>pseudo_eset, <span class="at">sc.eset.list =</span> pancreas.sc, <span class="at">ct.varname =</span> <span class="st">&quot;cluster&quot;</span>,</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>                                <span class="at">sample =</span> <span class="st">&quot;sample&quot;</span>, <span class="at">weight.basis =</span><span class="cn">TRUE</span>,<span class="at">truep =</span> truep, <span class="at">ct.sub =</span>  <span class="fu">c</span>(<span class="st">&quot;alpha&quot;</span>,<span class="st">&quot;beta&quot;</span>,<span class="st">&quot;delta&quot;</span>,<span class="st">&quot;gamma&quot;</span>), <span class="at">search.length =</span> <span class="fl">0.02</span>,<span class="at">grid.search=</span><span class="cn">TRUE</span>)</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>  comb<span class="ot">&lt;-</span>SCDC_results<span class="sc">$</span>prop.only</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>  weight_matrix<span class="ot">&lt;-</span>SCDC_results<span class="sc">$</span>w_table[<span class="st">&quot;mAD_Y&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>  SCDC_ENSEMBLE_MAD<span class="ot">&lt;-</span>SCDC<span class="sc">:::</span><span class="fu">wt_prop</span>(weight_matrix,comb)</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># saveRDS(SCDC_ENSEMBLE_MAD,&quot;/Users/chixiang.chen/Library/CloudStorage/OneDrive-UniversityofMarylandSchoolofMedicine/postdoc/postdoc/deconvolution/ref_based_rd/data_InteRD/SCDC_ENSEMBLE_MAD_baron.rds&quot;)</span></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>    <span class="co"># saveRDS(comb,&quot;/Users/chixiang.chen/Library/CloudStorage/OneDrive-UniversityofMarylandSchoolofMedicine/postdoc/postdoc/deconvolution/ref_based_rd/data_InteRD/comb_baron.rds&quot;)</span></span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co">#results based on InteRD1</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>SCDC_ENSEMBLE_MAD<span class="ot">&lt;-</span><span class="fu">readRDSFromWeb</span>(<span class="st">&quot;https://github.com/chencxxy28/Data/raw/main/data_InteRD/SCDC_ENSEMBLE_MAD_baron.rds&quot;</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>comb<span class="ot">&lt;-</span><span class="fu">readRDSFromWeb</span>(<span class="st">&quot;https://github.com/chencxxy28/Data/raw/main/data_InteRD/comb_baron.rds&quot;</span>)</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  list_marker<span class="ot">&lt;-</span><span class="fu">readRDSFromWeb</span>(<span class="st">&quot;https://github.com/chencxxy28/Data/raw/main/data_InteRD/list_markerbaron20.rds&quot;</span>) <span class="co">#get markers selected from xin et al (2016)</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  lambda_option<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">0.01</span>,<span class="fl">0.05</span>,<span class="fl">0.1</span>,<span class="dv">1</span>,<span class="dv">5</span>,<span class="dv">100</span>)</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  cell_type_unique<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">&quot;alpha&quot;</span>,<span class="st">&quot;beta&quot;</span>,<span class="st">&quot;delta&quot;</span>,<span class="st">&quot;gamma&quot;</span>)</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  InteRD1.output<span class="ot">&lt;-</span><span class="fu">InteRD1</span>(<span class="at">bulk.data =</span>pseudo.baron<span class="sc">$</span>pseudo_eset,list_marker,cell_type_unique,<span class="at">comb_used=</span>comb,lambda_option)</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>  InteRD1<span class="ot">&lt;-</span><span class="fu">InteRD.predict.prop</span>(<span class="at">InteRD.output=</span>InteRD1.output)</span></code></pre></div>
<p>After obtaining InteRD1, we can obtain InteRD2 by integrating prior
biological information into the deconvolution in a robust manner. The
prior information is the population-level mean cell-type proportions and
their corresponding standard deviations across samples that are obtained
from scRNA-seq data from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5228327/">Baron et
al.Â (2016)</a>.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>ave_est <span class="ot">=</span> <span class="fu">pop.ct.prop.scRNA</span>(<span class="at">scRNA=</span>baron[[<span class="st">&quot;sc.eset.qc&quot;</span>]],<span class="at">cell_type_unique=</span>cell_type_unique)<span class="sc">$</span>pop.ct.prop</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>ave_sd <span class="ot">=</span> <span class="fu">pop.ct.prop.scRNA</span>(<span class="at">scRNA=</span>baron[[<span class="st">&quot;sc.eset.qc&quot;</span>]],<span class="at">cell_type_unique=</span>cell_type_unique)<span class="sc">$</span>pop.ct.sd</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>lambda_option<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="fu">seq</span>(<span class="at">from=</span><span class="dv">1</span>,<span class="at">to=</span><span class="dv">20</span>,<span class="at">length=</span><span class="dv">4</span>),<span class="fu">seq</span>(<span class="at">from=</span><span class="dv">30</span>,<span class="at">to=</span><span class="dv">100</span>,<span class="at">length=</span><span class="dv">4</span>),<span class="dv">200</span>,<span class="dv">500</span>,<span class="dv">1000000</span><span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>InteRD2.output<span class="ot">&lt;-</span><span class="fu">InteRD2</span>(<span class="at">bulk.data=</span>pseudo.baron<span class="sc">$</span>pseudo_eset,list_marker,cell_type_unique,<span class="at">comb_sampled=</span>InteRD1,ave_est,ave_sd,<span class="at">lambda_option=</span>lambda_option)</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>InteRD2<span class="ot">&lt;-</span><span class="fu">InteRD.predict.prop</span>(<span class="at">InteRD.output=</span>InteRD2.output)</span></code></pre></div>
<p>Based on the evaluations, we can tell that InteRD-based estimates
(both InteRD1 and InteRD2) are better than the existing method.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">evaluate</span>(SCDC_ENSEMBLE_MAD,pseudo.baron<span class="sc">$</span>true_p)<span class="sc">$</span>all.eva</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">evaluate</span>(InteRD1,pseudo.baron<span class="sc">$</span>true_p)<span class="sc">$</span>all.eva</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="fu">evaluate</span>(InteRD2,pseudo.baron<span class="sc">$</span>true_p)<span class="sc">$</span>all.eva</span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
